src/tribler-core/tribler_core/modules/resource_monitor/core.py

```diff
@@ -0,0 +1,105 @@
+import os
+import time
+
+from ipv8.taskmanager import TaskManager
+
+import psutil
+
+from tribler_common.simpledefs import NTFY
+
+from tribler_core.modules.resource_monitor.base import ResourceMonitor
+
+FREE_DISK_THRESHOLD = 100 * (1024 * 1024)  # 100MB
+DEFAULT_RESOURCE_FILENAME = "resources.log"
+
+
+class CoreResourceMonitor(ResourceMonitor, TaskManager):
+    """
+    Implementation class of ResourceMonitor by the core process. The core process uses
+    TaskManager to implement start() and stop() methods.
+    """
+
+    def __init__(self, session):
+        TaskManager.__init__(self)
+        ResourceMonitor.__init__(self)
+        self.session = session
+
+        self.disk_usage_data = []
+
+        self.state_dir = session.config.get_state_dir()
+        self.resource_log_file = session.config.get_log_dir() / DEFAULT_RESOURCE_FILENAME
+        self.resource_log_enabled = session.config.get_resource_monitor_enabled()
+
+    def start(self):
+        """
+        Start the resource monitor by scheduling a LoopingCall.
+        """
+        self.register_task("check_resources", self.check_resources,
+                           interval=self.session.config.get_resource_monitor_poll_interval())
+
+    async def stop(self):
+        await self.shutdown_task_manager()
+        super().stop()
+
+    def check_resources(self):
+        ResourceMonitor.check_resources(self)
+        # Additionally, record the disk and notify on low disk space available.
+        self.record_disk_usage()
+
+        # Write resource logs
+        if self.resource_log_file:
+            self.write_resource_logs()
+
+    def write_resource_logs(self):
+        if not self.memory_data or not self.cpu_data:
+            return
+
+        if not self.resource_log_file.exists():
+            resource_dir = self.resource_log_file.parent
+            if not resource_dir.exists() and resource_dir:
+                os.makedirs(resource_dir)
+
+        with self.resource_log_file.open(mode="a+") as output_file:
+            latest_memory_data = self.memory_data[len(self.memory_data) - 1]
+            latest_cpu_data = self.cpu_data[len(self.memory_data) - 1]
```

Comment:
**issue:** both 63 and 64 lines can be a cause for the exception in case of `len() == 0`